DHT looping call churn

DanGraur opened the pull request:
This is an implementation of the enhancement suggested in #273.

DanGraur commented:
Ok, so I implemented what was suggested in the aforementioned issue. I initially moved all the behavior from the `ping_all` method to the new `PingChurn` class. However, during testing, I noticed that the `DHTDiscoveryCommunity` had its own `ping_all` method, which would call the `ping_all` in its superclass, i.e. in `DHTCommunity`. Since that wasn't there the behavior of `ping_all` in `DHTDiscoveryCommunity` was faulty, and hence could not pass a unit tests. Since calling the `PingChurn` strategy object stored in the `strategies` list of the `IPv8` object would imply a lot of bad practices, I decided to move `ping_all` back from `PingChurn` to `DHTCommunity` (`PingChurn` will now only call the method, rather than hold its entire logic). I suppose to a certain extent, this may improve reusability, as long as the base `overlay` has an implementation for `ping_all`. Is this ok, or should I find a way in which I actually move the entire behavior of `ping_all` to `take_step` in `PingChurn`? It should be mentioned, that there is no `LoopingCall` for `ping_all`, so the focus point of the ticket has been addressed.

qstokkink commented:
I like this as a first step, eventually we want to also move the innards of the current `ping_all` to the strategy.

This PR also shows me that we want to do the `configuration.py` differently. Overlays/communities should be able to expose strategy names (with corresponding classes) themselves, instead of having them globally known in `ipv8_service.py`. Your `PingChurn` is very much only for the DHT community. We also have the existing `RandomChurn` which can only be used by the DiscoveryCommunity.

qstokkink commented:
I moved my own suggestion to #354.

DanGraur commented:
If there's currently no changes left to be made to this branch, can this be accepted and merged (after I'll rebase)?

qstokkink commented on ipv8/test/mocking/community.py:
```diff
@@ -23,3 +24,21 @@ def __init__(self):
         self.last_bootstrap = 0
+
+
+class MockDHTCommunity(DHTCommunity):
```
It seems like this can be removed now.

qstokkink commented on ipv8/peerdiscovery/churn.py:
```diff
@@ -66,3 +66,22 @@ def take_step(self, service_id=None):
+        self.ping_interval = ping_interval
+
+    def take_step(self, service_id=None):
+        self.overlay.routing_table.remove_bad_nodes()
```
All DiscoveryStrategies should claim the `with self.walk_lock:` on `take_step`.

DanGraur commented:
I removed the `MockDHTCommunity` class, and added the `walk_lock` acquisition code in the `take_step` method of the `PingChurn` class. Also, I rebased this branch. Should I squash the commit history of this branch as well?
